### DISCUSSION TITLE
Support for NDI Output

### DISCUSSION THREAD

**Discussion by StanleyBonhomme** [praisenter/praisenter#37] [Ideas]
Add support for NDI output for lyrics & scriptures for live streams.

**Comment by wnbittle** [comment]
Thanks for the request - could you give some details on how you would like to see this feature added?  For example, would you want the NDI output to _mirror_ a standard output device (HDMI, DVI, VGA, etc.) that you are already showing?  Or would you want the NDI to be controlled separately (similar to how multiple standard output devices are done today)?

Here's some initial findings (mostly for myself):

- [A Java library that wraps the NewTek NDI libraries](https://github.com/WalkerKnapp/devolay)
- [An example of using that library to broadcast video by writing the pixels of every frame to the device](https://github.com/WalkerKnapp/devolay/blob/release/examples/src/main/java/me/walkerknapp/devolayexamples/SendVideoExample.java)
- JavaFX (the underlying technology that Praisenter uses for rendering) already supports rendering offscreen by creating a headless Scene and using the snapshot method to render to a WritableImage.

My main concern would be the last bullet point's performance.  That strategy would require JavaFX to render the output to a WritableImage (memory) first, then we would have to copy that to the NDI output buffer.  It _should_ be performant enough, but having JavaFX render directly to the NDI output buffer would be better.

**Comment by StanleyBonhomme** [comment]
What I had in mind was a way to send Scripture and Lyrics as transparency over NDI that way users can bring that content into there live streams. Like OBS and vMix

For example:

<img width="1673" alt="Screenshot 2021-01-01 at 15 19 38" src="https://github.com/praisenter/praisenter/assets/30869347/58a5fb3c-d695-41b2-9735-c256e2f34538">

**Comment by wnbittle** [comment]
What operating system are you on?  Windows, Ubuntu, Mac or Mac M1?

**Reply by StanleyBonhomme** [reply]
I am using windows 10

**Reply by wnbittle** [reply]
Thanks!  Are you only looking for the video stream or do you need the sound too?

What I mean is that Praisenter allows you to build a slide/template with images, text, video and audio.  In the case of Video and Audio, they may have an audio stream (sound).  When you connect OBS to the NDI output from Praisenter are you expecting to receive just the visual component or the sound component too?

**Comment by wnbittle** [comment]
Another finding.  So the design that I had originally thought of would only support Video.  Audio would not be supported since JavaFX doesn't allow access to the underlying audio stream.

**Comment by wnbittle** [comment]
Just a quick update.  I have this working for video only (no audio) on Windows 10.  I still need to clean some things up and expose to the user the ability to create a NDI output.  I also need to make sure I add the relevant attribution as required by the NDI SDK license.

**Comment by wnbittle** [comment]
@StanleyBonhomme there's a build ready that has NDI video support (no audio) if you want to try it out.

I still need to test on linux before publishing the change and a few more cleanup tasks.
